internal/impl/kafka/enterprise/kafka_migrator_input.go

```diff
@@ -0,0 +1,580 @@
+			}
+		}
+	}()
+
+	return nil
+}
+
+// ReadBatch attempts to extract a batch of messages from the target topics.
+func (r *KafkaMigratorReader) ReadBatch(ctx context.Context) (service.MessageBatch, service.AckFunc, error) {
+	r.connMut.Lock()
+	defer r.connMut.Unlock()
+
+	if r.client == nil {
+		return nil, nil, service.ErrNotConnected
+	}
+
+	// TODO: Is there a way to wait a while until we actually get f.batchSize messages instead of returning as many as
+	// we have right now? Otherwise, maybe switch back to `PollFetches()` and have `batch_byte_size` and `batch_period`
+	// via `FetchMinBytes`, `FetchMaxBytes` and `FetchMaxWait()`?
+
+	// TODO: Looks like when using `regexp_topics: true`, franz-go takes over a minute to discover topics which were
+	// created after `PollRecords()` was called for the first time. Might need to adjust the timeout for the internal
+	// topic cache.
+	fetches := r.client.PollRecords(ctx, r.batchSize)
+	if errs := fetches.Errors(); len(errs) > 0 {
+		// Any non-temporal error sets this true and we close the client
+		// forcing a reconnect.
+		nonTemporalErr := false
+
+		for _, kerr := range errs {
+			// TODO: The documentation from franz-go is top-tier, it
+			// should be straight forward to expand this to include more
+			// errors that are safe to disregard.
+			if errors.Is(kerr.Err, context.DeadlineExceeded) ||
+				errors.Is(kerr.Err, context.Canceled) {
+				continue
+			}
+
+			nonTemporalErr = true
+
+			if !errors.Is(kerr.Err, kgo.ErrClientClosed) {
+				r.mgr.Logger().Errorf("Kafka poll error on topic %v, partition %v: %v", kerr.Topic, kerr.Partition, kerr.Err)
+			}
+		}
+
+		if nonTemporalErr {
+			r.client.Close()
+			r.client = nil
+			return nil, nil, service.ErrNotConnected
+		}
+	}
+
+	// TODO: Is there a way to get the actual selected topics instead of all of them?
+	topics := r.client.GetConsumeTopics()
+	if r.regexPattern {
+		topics = slices.DeleteFunc(topics, func(topic string) bool {
+			for _, tp := range r.topicPatterns {
+				if tp.MatchString(topic) {
+					return false
+				}
+			}
+			return true
+		})
+	}
+
+	if len(topics) > 0 {
+		r.mgr.Logger().Debugf("Consuming from topics: %s", topics)
+	} else if r.regexPattern {
+		r.mgr.Logger().Warn("No matching topics found")
+	}
+
+	if !r.outputTopicsCreated {
+		var output *KafkaMigratorWriter
+		if res, ok := r.mgr.GetGeneric(r.topicSink); ok {
+			output = res.(*KafkaMigratorWriter)
+		} else {
+			r.mgr.Logger().Debugf("Writer for topic sink %q not found", r.topicSink)
+		}
+
+		if output != nil {
+			for _, topic := range topics {
+				r.mgr.Logger().Infof("Creating topic %q", topic)
+
+				if err := createTopic(ctx, topic, r.client, output.client); err != nil && err != errTopicAlreadyExists {
+					r.mgr.Logger().Errorf("Failed to create topic %q and ACLs: %s", topic, err)
+				} else {
+					if err == errTopicAlreadyExists {
+						r.mgr.Logger().Infof("Topic %q already exists", topic)
```

Comment:
This might be worth switching to debug level as I imagine it's noisy in some situations.

Reply:
Nice catch! ❤️